// pdv/index.js
console.log('load pdv ....');

exports.init = function(req, res){
	exports.res = res;
	exports.req = req;

	exports.testBox();

	var entrada = {v1 : req.query.v1, v2 : req.query.v2};
	console.log('ENTRADA '+JSON.stringify(entrada));
	exports.somaValores(entrada);
	
};

exports.somaValores = function(data) {	
	require('somaValores/index.js').init(data, function(args) {
		exports.convertMoney(args);	
	});
};

exports.convertMoney = function(data) {
	require('convertMoney/index.js').init(data, function(args) {
		console.log('SAÍDA '+JSON.stringify(args));
		exports.res.end('App (pdv) '+JSON.stringify(args));
	});
};

exports.testBox = function() {
	var tests = require('convertMoney/test.js');

	for(var i in tests){
			tests[i](function(result) {
				console.log('test ... '+JSON.stringify(result));
			});		
	}

	var tests = require('somaValores/test.js');

	for(var i in tests){
			tests[i](function(result) {
				console.log('test ... '+JSON.stringify(result));
			});		
	}
};

console.log('load pdv [ok]');
